Improved handling for disconnected network interfaces in network module

edward-shen opened the pull request:
This pull request updates the network module to support and handle non-existing interfaces, or interfaces that disappear after polybar is run. This situation occurs frequently when using a laptop that does not have an Ethernet port. USB-to-Ethernet interfaces are not guaranteed to exist during init and may exist after polybar has been initialized (#1164). Additionally, if the interface is connected during init but disappears after init, polybar will not update graphically (#1163).

Polybar currently will properly set `m_connected` to be false if it cannot properly query the interface. However, it does not update the displayed format. This PR resolves this by allowing the `update` function to continue even if it cannot query the interface correctly. It also implements the missing label substitutions for the `DISCONNECTED` state.

Polybar currently will attempt to create a reference to the network interface on initialization. This means that if an interface is specified but is not present during set up, polybar will hit a runtime exception and disable that module. This PR resolves this by handling the network error thrown by the `net` class in the `network` module, and attempts to set the interface on every update, if the interface reference does not exist yet.

patrick96 commented:
Thanks! Will review this in the coming days.

edward-shen commented:
Oh, I guess this also changes behavior with the config option, because it no longer will disable the module if an incorrect interface is named. I think this is acceptable behavior for modules that deal with disappearing and reappearing items, though.

edward-shen commented:
Hey, any updates on this PR?

patrick96 commented:
Hi, not yet unfortunately, been busy dealing with stuff on here. I have not forgotten about this though, it's on my TODO list

edward-shen commented:
No problem, take your time!

edward-shen commented on src/modules/network.cpp:
```diff
@@ -85,21 +85,20 @@ namespace modules {
         m_wireless ? static_cast<net::network*>(m_wireless.get()) : static_cast<net::network*>(m_wired.get());
 
```
Added.

edward-shen commented on src/modules/network.cpp:
```diff
@@ -133,13 +132,18 @@ namespace modules {
       replace_tokens(m_label[connection_state::DISCONNECTED]);
     }
 
+    if (m_label[connection_state::DISCONNECTED]) {
```
Was due to auto merge, sorry!

edward-shen commented:
My bad. I forgot that this PR was a thing and wanted to pull Polybar from upstream, and I somehow screwed it up.

I've forced pushed after resetting from hard and cherry picking those commits. The duplicated code was a auto-merge problem, so I've eliminated it. The two commits I've added now addresses your comments.

I've rebuilt this and made sure this works with a usb adapter.

Lomadriel commented:
> One thing that bothers me is that the two warnings `Failed to query interface` and `Could not access interface` are printed on every update. One way I could think of fixing this is by having a private `bool` that stores whether or not to display warnings, it it set to false once the warning is displayed and set to true once the update succeeds.
> I'm wondering whether or not there is a cleaner way to do this. What do you think?

I didn't pay attention to that but indeed, that definitively better for the end-user.
I do not see a better solution than a bool for implementing this but that seems correct.

tremby commented:
It'd be nice to be able to differentiate between absent interface (e.g. don't show the network status module at all) and present interface with no network connection (e.g. show that ethernet link is down). Any chance of allowing that?

doronbehar commented:
Thank you @edward-shen for writing this patch :) I can confirm this patch works fine for me with an almost-never-used usb tethering interface `usb0`.
